src/shipit_code_coverage_backend/shipit_code_coverage_backend/v2/diff.py

```diff
@@ -0,0 +1,200 @@
+import asyncio
+import itertools
+
+import aiohttp
+
+from cli_common import log
+from shipit_code_coverage_backend.services.active_data import ActiveDataCoverage
+from shipit_code_coverage_backend.v2.base import NoResults
+from shipit_code_coverage_backend.v2.base import active_data
+from shipit_code_coverage_backend.v2.base import active_data_async
+
+logger = log.get_logger(__name__)
+
+
+async def coverage_in_push(files, push):
+    '''
+    Load coverage for several files, on a specific push
+    Aggregate covered lines directly through ES "painless"
+    '''
+    assert isinstance(files, (list, tuple))
+    assert isinstance(push, int)
+
+    filters = [
+        # Filter by files
+        {'terms': {ActiveDataCoverage.FIELD_FILENAME: files}},
+
+        # Filter by push
+        {'term': {ActiveDataCoverage.FIELD_PUSH: push}},
+    ]
+    query = ActiveDataCoverage.base_query(filters)
+    query.update({
+        'size': 0,
+        'aggs': {
+            'files': {
+                'terms': {'field': ActiveDataCoverage.FIELD_FILENAME},
+
+                'aggs': {
+                    'covered': {
+                        'scripted_metric': {
+                            'init_script': 'params._agg.covered = []',
+
+                            # Merge all the covered lines, per ES shard
+                            'map_script': "params._agg.covered.addAll(doc['source.file.covered.~n~'])",
+
+                            # Make a list with all items per shard
+                            'combine_script': 'return params._agg.covered.stream().collect(Collectors.toList())',
+
+                            # Merge all shards lists into a sorted set per file (top aggregation)
+                            'reduce_script': 'return params._aggs.stream().flatMap(Collection::stream).sorted().distinct().collect(Collectors.toList())'
+                        }
+                    }
+                }
+            },
+        },
+    })
+    try:
+        out = await active_data_async.search('coverage_files_push', query, timeout=100)
+    except NoResults:
+        return {}
+
+    return {
+        bucket['key']: list(map(int, bucket['covered']['value']))
+        for bucket in out['aggregations']['files']['buckets']
+    }
+
+
+async def changes_on_files(files, push, revision_index, branch_name):
+    '''
+    List all the operations that happened on some files, in a push
+    '''
+    assert isinstance(files, (list, tuple))
+    assert isinstance(push, int)
+    query = {
+        'query': {
+            'bool': {
+
+                # Files intersection
+                'should': [
+                    {'match': {'changeset.files': f}}
+                    for f in files
+                ],
+
+                'must': [
+                    # After our commit
+                    {'range': {
+                        'index': {'gt': revision_index},
+                    }},
+
+                    # Same push and repo
```

Comment:
It's not always the same push.
Example:
commit A
commit B
push 1

commit C
commit D
push 2

If the coverage build fails on push 1 for some reason, we have to load the coverage data from push 2.